### PR TITLE
CY-1507 Allow get_input and get_property at runtime

### PR DESCRIPTION
**Pull request by tehasdf**

### CODE REVIEW

**Comment by tehasdf** [comment]
This heavily depends on cloudify-cosmo/cloudify-common#356 and that should be reviewed first

File: rest-service/manager_rest/storage/resource_models.py [hunk 1]
```diff
@@ -563,14 +563,14 @@ class Node(SQLResourceBase):
...
 
-    deploy_number_of_instances = db.Column(db.Integer, nullable=False)
     # TODO: This probably should be a foreign key, but there's no guarantee
     # in the code, currently, that the host will be created beforehand
     host_id = db.Column(db.Text)
-    max_number_of_instances = db.Column(db.Integer, nullable=False)
-    min_number_of_instances = db.Column(db.Integer, nullable=False)
-    number_of_instances = db.Column(db.Integer, nullable=False)
-    planned_number_of_instances = db.Column(db.Integer, nullable=False)
+    deploy_number_of_instances = db.Column(JSONString, nullable=False)
```
Comment (tehasdf) [inline rest-service/manager_rest/storage/resource_models.py]: TODO migrations

File: tests/integration_tests/tests/agentless_tests/test_deployment_modification.py [hunk 2]
```diff
@@ -307,3 +307,34 @@ def test_group_deployment_modification(self):
         scaling_group = deployment['scaling_groups']['compute_and_ip']
         self.assertEqual(2, scaling_group['properties']['planned_instances'])
         self.assertEqual(2, scaling_group['properties']['current_instances'])
+
+    def test_deployment_modification_functions(self):
```
Comment (tehasdf) [inline tests/integration_tests/tests/agentless_tests/test_deployment_modification.py]: TODO a few more tests, see comments under CY-1507

File: rest-service/manager_rest/dsl_functions.py [hunk 3]
```diff
@@ -32,86 +32,106 @@
...
     sm = get_storage_manager()
-    methods = _get_methods(None, sm)
-    return methods['get_secret_method']
+    secret = sm.get(Secret, secret_key)
+    decrypted_value = cryptography_utils.decrypt(secret.value)
+    return SecretType(secret_key, decrypted_value)
 
 
-def _get_methods(deployment_id, storage_manager):
+class DBEvaluationStorage(object):
```
Comment (tehasdf) [inline rest-service/manager_rest/dsl_functions.py]: TODO check performance of this, most likely needs more prefetch

File: tests/integration_tests/tests/agentless_tests/test_runtime_functions.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+#    * See the License for the specific language governing permissions and
+#    * limitations under the License.
+
+import os
+
+from integration_tests import AgentlessTestCase
+from integration_tests.tests.utils import get_resource as resource
+
+
+class TestRuntimeFunctionEvaluation(AgentlessTestCase):
```
Comment (tehasdf) [inline tests/integration_tests/tests/agentless_tests/test_runtime_functions.py]: TODO comments for the test and test blueprint

File: rest-service/manager_rest/dsl_functions.py [hunk 5]
```diff
@@ -32,86 +32,113 @@
 SecretType = namedtuple('Secret', 'key value')
 
 
+def evaluate_node(node):
+    node['name'] = node['id']
+    deployment_id = node['deployment_id']
+    sm = get_storage_manager()
+    sm.get(Deployment, deployment_id, include=['id'])
```
Comment (tyacbovi) [inline rest-service/manager_rest/dsl_functions.py]: Why is this needed? for validating the deployment_id is for a "live" deployment?

File: rest-service/manager_rest/dsl_functions.py [hunk 6]
```diff
@@ -32,86 +32,113 @@
 SecretType = namedtuple('Secret', 'key value')
 
 
+def evaluate_node(node):
+    node['name'] = node['id']
```
Comment (tyacbovi) [inline rest-service/manager_rest/dsl_functions.py]: why do we need this? did you switch to use id?

File: rest-service/manager_rest/dsl_functions.py [hunk 7]
```diff
@@ -32,86 +32,113 @@
...
+    deployment_id = node['deployment_id']
+    sm = get_storage_manager()
+    sm.get(Deployment, deployment_id, include=['id'])
+    storage = DBEvaluationStorage(deployment_id, sm)
+    try:
+        return functions.evaluate_node_functions(node, storage)
+    except parser_exceptions.FunctionEvaluationError, e:
```
Comment (tyacbovi) [inline rest-service/manager_rest/dsl_functions.py]: , => as

File: rest-service/manager_rest/rest/resources_v3/nodes.py [hunk 8]
```diff
@@ -14,25 +14,26 @@
...
+    evaluate_node_instance
+)
 
 from .. import rest_decorators
 from ..resources_v1.nodes import NodeInstancesId as v1_NodeInstancesId
 from ..resources_v2 import Nodes as v2_Nodes
 
 
 class Nodes(v2_Nodes):
-    @rest_decorators.exceptions_handled
```
Comment (tyacbovi) [inline rest-service/manager_rest/rest/resources_v3/nodes.py]: Why is this needed?

File: rest-service/manager_rest/rest/resources_v3/nodes.py [hunk 9]
```diff
@@ -14,25 +14,26 @@
...
     @authorize('node_list')
     @rest_decorators.evaluate_functions
-    def get(self, evaluate_functions=False, *args, **kwargs):
+    def get(self, evaluate_functions=True, *args, **kwargs):
```
Comment (tyacbovi) [inline rest-service/manager_rest/rest/resources_v3/nodes.py]: so this will be set with the eager flag, right?

File: rest-service/manager_rest/dsl_functions.py [hunk 10]
```diff
@@ -32,86 +32,113 @@
...
     sm = get_storage_manager()
-    methods = _get_methods(None, sm)
-    return methods['get_secret_method']
+    secret = sm.get(Secret, secret_key)
+    decrypted_value = cryptography_utils.decrypt(secret.value)
+    return SecretType(secret_key, decrypted_value)
 
 
-def _get_methods(deployment_id, storage_manager):
+class DBEvaluationStorage(object):
```
Comment (tyacbovi) [inline rest-service/manager_rest/dsl_functions.py]: maybe DBEvaluationStorage => EvalutedFunctionsStorage

File: tests/integration_tests/tests/agentless_tests/test_runtime_functions.py [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+            dep_update.id)
+
+    def test_manual_update_functions(self):
+        deployment = self.deploy(
+            self.bp_path,
+            inputs={'input1': self.BASE_VALUE},
+            functions_eager=False)
+        self.execute_workflow('install', deployment.id)
+        self._assert_properties(deployment, self.BASE_VALUE)
+        self._manual_deployment_update(deployment)
```
Comment (tyacbovi) [inline tests/integration_tests/tests/agentless_tests/test_runtime_functions.py]: why is this needed as an external script?

File: tests/integration_tests/tests/agentless_tests/test_runtime_functions.py [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+        instances_node1 = self.client.node_instances.list(
+            node_id='node1', deployment_id=deployment.id)
+        self.assertEqual(len(instances_node1), 1)
+        instances_node2 = self.client.node_instances.list(
+            node_id='node2', deployment_id=deployment.id)
+        ni1 = instances_node1[0]
+        ni2 = instances_node2[0]
+
+        outputs = self.client.deployments.outputs.get(deployment.id)['outputs']
+        self.assertEqual(outputs['out1'], value)
```
Comment (tyacbovi) [inline tests/integration_tests/tests/agentless_tests/test_runtime_functions.py]: well it can make you think that it always will return value

File: tests/integration_tests/tests/agentless_tests/test_runtime_functions.py [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+            deployment_id=deployment.id,
+            blueprint_or_archive_path=new_bp.name,
+            inputs={'input1': self.CHANGED_VALUE, 'fail_create': False}
+        )
+        execution = self.client.executions.get(dep_update.execution_id)
+        self.wait_for_execution_to_end(execution)
+        dep_update = self.client.deployment_updates.finalize_commit(
+            dep_update.id)
+
+    def test_manual_update_functions(self):
```
Comment (tyacbovi) [inline tests/integration_tests/tests/agentless_tests/test_runtime_functions.py]: maybe => test_manual_update_functions_values_updates

File: rest-service/manager_rest/storage/resource_models.py [hunk 14]
```diff
@@ -565,14 +565,14 @@ class Node(SQLResourceBase):
...
```
Comment (geokala) [inline rest-service/manager_rest/storage/resource_models.py]: Probably put a remark here that the columns have to be strings because of potential get_inputs?

(but then does the number_of_instances and planned_number_of_instances ever manage to not be an int?)

File: tests/integration_tests/tests/agentless_tests/test_runtime_functions.py (hunk 13 above)
Comment (geokala) [reply tests/integration_tests/tests/agentless_tests/test_runtime_functions.py]: If it were to be changed, test_manual_update_functions_update_values would parse better.

File: rest-service/manager_rest/dsl_functions.py (hunk 6 above)
Comment (tehasdf) [reply rest-service/manager_rest/dsl_functions.py]: yes, dsl-parser uses 'name', while the storage uses 'id'. Added a comment saying that

File: rest-service/manager_rest/dsl_functions.py (hunk 5 above)
Comment (tehasdf) [reply rest-service/manager_rest/dsl_functions.py]: yes i think that is why (this isn't a new line, just copied)

File: rest-service/manager_rest/dsl_functions.py (hunk 10 above)
Comment (tehasdf) [reply rest-service/manager_rest/dsl_functions.py]: `FunctionEvaluationDBStorage`, it should have `DB` to distinguish from other function evaluation storages (local's FileStorage and InMemoryStorage)

File: rest-service/manager_rest/rest/resources_v3/nodes.py (hunk 8 above)
Comment (tehasdf) [reply rest-service/manager_rest/rest/resources_v3/nodes.py]: Committed by accident

our `rest_decorators.exceptions_handled` is absolutely fucked and you need to remove it if you want to see exceptions

made CY-1687 about it

File: tests/integration_tests/tests/agentless_tests/test_runtime_functions.py (hunk 12 above)
Comment (tehasdf) [reply tests/integration_tests/tests/agentless_tests/test_runtime_functions.py]: please reword this comment so that i can understand

File: tests/integration_tests/tests/agentless_tests/test_runtime_functions.py (hunk 11 above)
Comment (tehasdf) [reply tests/integration_tests/tests/agentless_tests/test_runtime_functions.py]: because it must run on docl, because it uses storage-manager

File: tests/integration_tests/tests/agentless_tests/test_runtime_functions.py (hunk 13 above)
Comment (tehasdf) [reply tests/integration_tests/tests/agentless_tests/test_runtime_functions.py]: changed the test names to `test_(manual_update/deployment_update)_(new_workflow/resume)`

File: packaging/mgmtworker/requirements.txt [hunk 15]
```diff
@@ -1,5 +1,5 @@
-git+https://github.com/cloudify-cosmo/cloudify-common@master
-git+https://github.com/cloudify-cosmo/cloudify-agent@master#egg=cloudify-agent[celery,kerberos]==master
+git+https://github.com/cloudify-cosmo/cloudify-common@cy-1507-functions-runtime-1
```
Comment (tehasdf) [inline packaging/mgmtworker/requirements.txt]: remember

File: tests/integration_tests/tests/agentless_tests/test_runtime_functions.py (hunk 12 above)
Comment (tyacbovi) [reply tests/integration_tests/tests/agentless_tests/test_runtime_functions.py]: Sorry :) I meant that because you always return 'value' how can we be sure that it can return more than one value

File: tests/integration_tests/tests/agentless_tests/test_runtime_functions.py [hunk 16]
```diff
@@ -0,0 +1,197 @@
...
+        self.assertEqual(ni1.runtime_properties['prop1_value'], value)
+        self.assertEqual(ni1.runtime_properties['prop2_value'], value)
+        self.assertEqual(ni1.runtime_properties['source_input_value'], value)
+        self.assertEqual(ni1.runtime_properties['target_input_value'], value)
+        self.assertEqual(ni2.runtime_properties['source_input_value'], value)
+        self.assertEqual(ni2.runtime_properties['target_input_value'], value)
+        self.assertEqual(ni1.runtime_properties['source_prop_value'], value)
+        self.assertEqual(ni2.runtime_properties['target_prop_value'], value)
+
+    def _manual_deployment_update(self, deployment):
```
Comment (tyacbovi) [inline tests/integration_tests/tests/agentless_tests/test_runtime_functions.py]: so this is done like this for saving test time?

File: rest-service/manager_rest/deployment_update/handlers.py [hunk 17]
```diff
@@ -184,6 +186,43 @@ def modify(self, ctx, current_entities):
...
+    def _update_stored_operations(self, ctx, node, new_operation):
+        """Update the operations table with the new operation inputs."""
+
+        def _filter_operation(column):
+            # path in the parameters dict that stores the node name
+            node_name_path = ('task_kwargs', 'cloudify_context', 'node_name')
+            # this will use postgres' json operators
+            return cast(column, JSON)[node_name_path].astext == node.id
+
+        if not ctx.deployment_update.update_executions:
```
Comment (tyacbovi) [inline rest-service/manager_rest/deployment_update/handlers.py]: I am not sure that 'update_executions' is really clear, because if don't know what it does you can think that it could do more.

Comment (tyacbovi) [inline rest-service/manager_rest/deployment_update/handlers.py]: 'update_executions' does the user choose this? if yes integ tests are missing

File: tests/integration_tests/tests/agentless_tests/test_runtime_functions.py (hunk 12 above)
Comment (tehasdf) [reply tests/integration_tests/tests/agentless_tests/test_runtime_functions.py]: deployment outputs tests check that deployment outputs can return different values

File: tests/integration_tests/tests/agentless_tests/test_runtime_functions.py (hunk 16 above)
Comment (tehasdf) [reply tests/integration_tests/tests/agentless_tests/test_runtime_functions.py]: this is pretty much mocking out deployment update (if you wanna be technical it's more of a fake than a mock i guess...)
to check that the runtime-functions stuff is independent of deployment updates (because maybe all the work was done by the update?)

Comment (tehasdf) [reply tests/integration_tests/tests/agentless_tests/test_runtime_functions.py]: added a note about this to docstring
